File: src/main/java/lambda/evaluator/NodeGenerator.java
package lambda.evaluator;

import lambda.core.datamodels.AtomNode;
import lambda.core.datamodels.ExpressionListNode;
import lambda.core.datamodels.Node;

import java.util.List;

interface NodeGenerator {

    AtomNode generateAtomNode(String value);

    ExpressionListNode generateExpressionListNode(List<Node> children);

    static NodeGenerator newInstance() {
        return new NodeGenerator() {
            @Override
            public AtomNode generateAtomNode(String value) {
                return new AtomNode(value);
            }

            @Override
            public ExpressionListNode generateExpressionListNode(List<Node> children) {
                return new ExpressionListNode(children);
            }
        };
    }
}

File: src/main/java/lambda/evaluator/NodeGenerator.kt
package lambda.evaluator

import lambda.core.datamodels.*

internal interface NodeGenerator {
    fun generateAtomNode(
        value: String
    ): AtomNode

    fun generateExpressionListNode(
        children: List<Node>
    ): ExpressionListNode
}

internal class NodeGeneratorImpl: NodeGenerator {

    override fun generateAtomNode(
        value: String
    ): AtomNode {
        return AtomNode(value)
    }

    override fun generateExpressionListNode(
        children: List<Node>
    ): ExpressionListNode {
        return ExpressionListNode(
            children
        )
    }
}